app/controllers/actions/MessageTransformer.scala

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2021 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers.actions
+
+import com.google.inject.ImplementedBy
+import com.google.inject.Inject
+import logging.Logging
+import models.ArrivalRejectedResponse
+import models.ChannelType
+import models.GoodsReleasedResponse
+import models.Message
+import models.MessageResponse
+import models.MessageType
+import models.UnloadingPermissionResponse
+import models.UnloadingRemarksRejectedResponse
+import models.UnloadingRemarksResponse
+import models.XMLSubmissionNegativeAcknowledgementResponse
+import models.request.ArrivalRequest
+import play.api.mvc.Results.BadRequest
+import play.api.mvc._
+import play.twirl.api.HtmlFormat
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import scala.xml.NodeSeq
+
+class MessageTransformer @Inject()(implicit val executionContext: ExecutionContext) extends MessageTransformerInterface with Logging {
+
+  override def refine[A](request: ArrivalRequest[A]): Future[Either[Result, MessageTransformRequest[A]]] =
+    request.body match {
+      case x: NodeSeq =>
+        x.headOption.flatMap(node => messageResponse(request.channel)(node.label)) match {
```

Comment:
Might be worth double checking that the head node will always be the identifier. Not sure if api adds any wrappers etc.